deepmedic/dataManagement/preprocessing.py

```diff
@@ -10,58 +10,195 @@
-
-
-#In the 3 first axes. Which means it can take a 4-dim image.
-def unpadCnnOutputs(array1, tupleOfPaddingPerAxesLeftRight) :
-    #tupleOfPaddingPerAxesLeftRight : ( (padLeftR, padRightR), (padLeftC,padRightC), (padLeftZ,padRightZ)).
-    unpaddedArray1 = array1[tupleOfPaddingPerAxesLeftRight[0][0]:, tupleOfPaddingPerAxesLeftRight[1][0]:, tupleOfPaddingPerAxesLeftRight[2][0]:]
-    #The checks below are to make it work if padding == 0, which may happen for 2D on the 3rd axis.
-    unpaddedArray1 = unpaddedArray1[:-tupleOfPaddingPerAxesLeftRight[0][1],:,:] if tupleOfPaddingPerAxesLeftRight[0][1] > 0 else unpaddedArray1 
-    unpaddedArray1 = unpaddedArray1[:,:-tupleOfPaddingPerAxesLeftRight[1][1],:] if tupleOfPaddingPerAxesLeftRight[1][1] > 0 else unpaddedArray1
-    unpaddedArray1 = unpaddedArray1[:,:,:-tupleOfPaddingPerAxesLeftRight[2][1]] if tupleOfPaddingPerAxesLeftRight[2][1] > 0 else unpaddedArray1
+    tupleOfPaddingPerAxes = ((paddingAtLeftPerAxis[0], paddingAtRightPerAxis[0]),
+                             (paddingAtLeftPerAxis[1], paddingAtRightPerAxis[1]),
+                             (paddingAtLeftPerAxis[2], paddingAtRightPerAxis[2]))
+    # Very poor design because channels/gt/bmask etc are all getting back a different padding?
+    # tupleOfPaddingPerAxes is returned in order for unpad to know.
+    return [np.lib.pad(array1, tupleOfPaddingPerAxes, 'reflect'), tupleOfPaddingPerAxes]
+
+
+# In the 3 first axes. Which means it can take a 4-dim image.
+def unpadCnnOutputs(array1, tupleOfPaddingPerAxesLeftRight):
+    # tupleOfPaddingPerAxesLeftRight : ( (padLeftR, padRightR), (padLeftC,padRightC), (padLeftZ,padRightZ)).
+    unpaddedArray1 = array1[tupleOfPaddingPerAxesLeftRight[0][0]:,
+                     tupleOfPaddingPerAxesLeftRight[1][0]:,
+                     tupleOfPaddingPerAxesLeftRight[2][0]:]
+    # The checks below are to make it work if padding == 0, which may happen for 2D on the 3rd axis.
+    unpaddedArray1 = unpaddedArray1[:-tupleOfPaddingPerAxesLeftRight[0][1], :, :] \
+        if tupleOfPaddingPerAxesLeftRight[0][1] > 0 else unpaddedArray1
+    unpaddedArray1 = unpaddedArray1[:, :-tupleOfPaddingPerAxesLeftRight[1][1], :] \
+        if tupleOfPaddingPerAxesLeftRight[1][1] > 0 else unpaddedArray1
+    unpaddedArray1 = unpaddedArray1[:, :, :-tupleOfPaddingPerAxesLeftRight[2][1]] \
+        if tupleOfPaddingPerAxesLeftRight[2][1] > 0 else unpaddedArray1
+
     return unpaddedArray1
 
-    
 
+def neg_val_check(img, log):
+    is_neg_int = img < 0
+    num_voxels_neg_int = np.sum(is_neg_int)
+    if num_voxels_neg_int > 0:
+        log.print3("[WARNING] One image has " + str(num_voxels_neg_int) +
+                   " voxels with intensity below zero! Unexpected!\n" +
+                   "The voxels with negative ints have (min, max, mean) = (" + str(np.min(img[is_neg_int])) + ", " +
+                   str(np.max(img[is_neg_int])) + ", " + str(np.mean(img[is_neg_int])) + ").")
+
+
```

Comment:
Likely we should make a Pre-Processor (or normalizer?) class and Z-score-norm should be a child?

Reply:
agree